subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c

```diff
@@ -0,0 +1,171 @@
+}
+
+static uint32_t idx_cut(uint32_t len, uint32_t idx)
+{
+	/* It is implicitly assumed a and b cannot differ by more then len. */
+	return (idx >= len) ? (idx - len) : (idx);
+}
+
+struct icmsg_buf *icmsg_buf_init(void *buf, size_t blen)
+{
+	/* blen must be big enough to contain icmsg_buf struct, byte of data
+	 * and message len (2 bytes).
+	 */
+	struct icmsg_buf *ib = buf;
+
+	__ASSERT_NO_MSG(blen > (sizeof(*ib) + sizeof(uint16_t)));
+
+	ib->len = blen - sizeof(*ib);
+	ib->wr_idx = 0;
+	ib->rd_idx = 0;
+
+	__sync_synchronize();
+	sys_cache_data_range(ib, sizeof(*ib), K_CACHE_WB);
+
+	return ib;
+}
+
+int icmsg_buf_write(struct icmsg_buf *ib, const char *buf, uint16_t len)
+{
+	/* The length of buffer is immutable - avoid reloading that may happen
+	 * due to memory bariers.
+	 */
+	const uint32_t iblen = ib->len;
+
+	/* rx_idx == wr_idx means the buffer is empty.
+	 * Max bytes that can be stored is len - 1.
+	 */
+	const uint32_t max_len = iblen - 1;
+
+	sys_cache_data_range(ib, sizeof(*ib), K_CACHE_INVD);
+	__sync_synchronize();
+
+	uint32_t wr_idx = ib->wr_idx;
+	uint32_t rd_idx = ib->rd_idx;
+
+	if ((len + sizeof(len) > max_len) || (len == 0)) {
+		/* Incorrect call. */
+		return -EINVAL;
+	}
+
+	uint32_t avail = max_len - idx_occupied(iblen, wr_idx, rd_idx);
+
+	if (len + sizeof(len) > avail) {
+		/* No free space. */
+		return -ENOMEM;
+	}
+
+	/* Store info about the message length. */
+	ib->data[wr_idx] = (uint8_t)len;
+	wr_idx = idx_cut(iblen, wr_idx + sizeof(uint8_t));
+	ib->data[wr_idx] = (uint8_t)(len >> 8);
+	wr_idx = idx_cut(iblen, wr_idx + sizeof(uint8_t));
+
+	/* Write until the end of the buffer. */
+	uint32_t sz = MIN(len, iblen - wr_idx);
+
+	memcpy(&ib->data[wr_idx], buf, sz);
+	__sync_synchronize();
+	sys_cache_data_range(&ib->data[wr_idx], sz, K_CACHE_WB);
+
+	if (len > sz) {
+		/* Write remaining data at the buffer head. */
+		memcpy(&ib->data[0], buf + sz, len - sz);
+		__sync_synchronize();
+		sys_cache_data_range(&ib->data[0], len - sz, K_CACHE_WB);
+	}
+
+	/* Update write index - make other side aware data was written. */
+	wr_idx = idx_cut(iblen, wr_idx + len);
+	ib->wr_idx = wr_idx;
+
+	__sync_synchronize();
+	sys_cache_data_range(ib, sizeof(*ib), K_CACHE_WB);
+
+	return len;
+}
+
+int icmsg_buf_read(struct icmsg_buf *ib, char *buf, uint16_t len)
+{
+	/* The length of buffer is immutable - avoid reloading. */
+	const uint32_t iblen = ib->len;
+
+	sys_cache_data_range(ib, sizeof(*ib), K_CACHE_INVD);
+	__sync_synchronize();
+
+	uint32_t rd_idx = ib->rd_idx;
+	uint32_t wr_idx = ib->wr_idx;
+	/* Temp rd_idx used to read message len. */
+	uint32_t rd_idx_temp = rd_idx;
+
+	if (rd_idx == wr_idx) {
+		/* The buffer is empty. */
+		return 0;
+	}
+
+	sys_cache_data_range(&ib->data[rd_idx], sizeof(ib->data[rd_idx]), K_CACHE_INVD);
+	__sync_synchronize();
+
+	/* Read message len. */
+	uint16_t mlen = ib->data[rd_idx_temp];
+	rd_idx_temp = idx_cut(iblen, wr_idx + sizeof(uint8_t));
+	mlen += ib->data[rd_idx_temp] >> 8;
```

Comment:
Please use `or` instear of `add`. It cannot be slower and usually is faster. It's also what we need here. I also recommend to add brackets.

`mlen |= (ib->data[rd_idx_temp] >> 8)`